python/arcticdb/storage_fixtures/s3.py

```diff
@@ -0,0 +1,275 @@
+            dest.copy({"Bucket": self.bucket, "Key": key}, key, SourceClient=source_client)
+
+
+class BaseS3StorageFixtureFactory(StorageFixtureFactory):
+    """Logic and fields common to real and mock S3"""
+
+    endpoint: str
+    region: str
+    default_key: Key
+    default_bucket: Optional[str] = None
+    default_prefix: Optional[str] = None
+    clean_bucket_on_fixture_exit = True
+
+    def __str__(self):
+        return f"{type(self).__name__}[{self.default_bucket or self.endpoint}]"
+
+    @staticmethod
+    def from_environment_variables():
+        out = BaseS3StorageFixtureFactory()
+        out.endpoint = os.getenv("ARCTICDB_REAL_S3_ENDPOINT")
+        out.region = os.getenv("ARCTICDB_REAL_S3_REGION")
+        out.default_bucket = os.getenv("ARCTICDB_REAL_S3_BUCKET")
+        access_key = os.getenv("ARCTICDB_REAL_S3_ACCESS_KEY")
+        secret_key = os.getenv("ARCTICDB_REAL_S3_SECRET_KEY")
+        out.default_key = Key(access_key, secret_key, "unknown user")
+        out.clean_bucket_on_fixture_exit = os.getenv("ARCTICDB_REAL_S3_CLEAR").lower() in ["true", "1"]
+        return out
+
+    def _boto(self, service: str, key: Key, api="client"):
+        import boto3
+
+        ctor = getattr(boto3, api)
+        return ctor(
+            service_name=service,
+            endpoint_url=self.endpoint if service == "s3" else self._iam_endpoint,
+            region_name=self.region,
+            aws_access_key_id=key.id,
+            aws_secret_access_key=key.secret,
+        )
+
+    def create_fixture(self) -> S3Bucket:
+        return S3Bucket(self)
+
+    def cleanup_bucket(self, b: S3Bucket):
+        # When dealing with a potentially shared bucket, we only clear our the libs we know about:
+        b.slow_cleanup(failure_consequence="We will be charged unless we manually delete it. ")
+
+
+class MotoS3StorageFixtureFactory(BaseS3StorageFixtureFactory):
+    default_key = Key("awd", "awd", "dummy")
+    _RO_POLICY: str
+    _RW_POLICY: str
+    host = "localhost"
+    region = "us-east-1"
+    port: int
+    endpoint: str
+    _enforcing_permissions = False
+    _iam_endpoint: str
+    _p: multiprocessing.Process
+    _s3_admin: Any
+    _iam_admin: Any = None
+    _bucket_id = 0
+    _live_buckets: List[S3Bucket] = []
+
+    @staticmethod
+    def run_server(port):
+        import werkzeug
+        from moto.server import DomainDispatcherApplication, create_backend_app
+
+        class _HostDispatcherApplication(DomainDispatcherApplication):
+            """The stand-alone server needs a way to distinguish between S3 and IAM. We use the host for that"""
+
+            _MAP = {"localhost": "s3", "127.0.0.1": "iam"}
+
+            _reqs_till_rate_limit = -1
+
+            def get_backend_for_host(self, host):
+                return self._MAP.get(host, host)
+
+            def __call__(self, environ, start_response):
+                path_info: bytes = environ.get("PATH_INFO", "")
+
+                with self.lock:
```

Comment:
I don't think we need this feature, since we already have the storage failure simulator in the C++ layer, and I'm reluctant for us to have multiple storage failure simulation mechanisms.

Reply:
I can take the commit out.

However, I don't think testing the storage exceptions solely in C++ is sufficient because it's not easy to test if the correct exception type is bubbled up to Python.

Reply:
The failure simulator has a way to configure itself with a protobuf object from the Python layer, allowing for an end to end test.

Reply:
I meant is it testing pybind as well?

Reply:
Yes

Reply:
OK after our big chat right now I think this is in fact the way to go, since we don't have a good stub S3 client implementation at the moment. Let's keep this feature in